include/picongpu/plugins/openPMD/Json.cpp

```diff
@@ -153,38 +153,52 @@ namespace
         return res;
     }
 
-    void readPattern(std::vector<Pattern>& patterns, nlohmann::json const& object)
+    KindOfConfig readPattern(
+        std::vector<Pattern>& patterns,
+        nlohmann::json& defaultConfig,
+        nlohmann::json const& object)
     {
         static std::string const errorMsg = R"END(
 [openPMD plugin] Each single pattern in an extended JSON configuration
 must be a JSON object with keys 'select' and 'cfg'.
-The key 'select' must point to either a single string or an array of strings.)END"
-                                            "";
+The key 'select' is optional, indicating a default configuration if it is
+not set.
+The key 'select' must point to either a single string or an array of strings.)END";
 
         if(!object.is_object())
         {
             throw std::runtime_error(errorMsg);
         }
         try
         {
-            nlohmann::json const& pattern = object.at("select");
             nlohmann::json const& cfg = object.at("cfg");
-            auto cfgShared = std::make_shared<nlohmann::json>(cfg);
-            if(pattern.is_string())
+            if(!object.contains("select"))
             {
-                patterns.emplace_back(pattern.get<std::string>(), std::move(cfgShared));
+                nlohmann::json const& cfg = object.at("cfg");
+                defaultConfig = cfg;
+                return KindOfConfig::Default;
             }
-            else if(pattern.is_array())
+            else
             {
-                patterns.reserve(pattern.size());
-                for(size_t i = 0; i < pattern.size(); ++i)
+                nlohmann::json const& pattern = object.at("select");
+                auto cfgShared = std::make_shared<nlohmann::json>(cfg);
+                if(pattern.is_string())
                 {
-                    patterns.emplace_back(pattern[i].get<std::string>(), cfgShared);
+                    patterns.emplace_back(pattern.get<std::string>(), std::move(cfgShared));
                 }
-            }
-            else
-            {
-                throw std::runtime_error(errorMsg);
+                else if(pattern.is_array())
+                {
+                    patterns.reserve(pattern.size());
+                    for(size_t i = 0; i < pattern.size(); ++i)
+                    {
+                        patterns.emplace_back(pattern[i].get<std::string>(), cfgShared);
```

Comment:
Why is `cfgShared` moved in the first branch, but not in the second?

Reply:
In the first branch, it is only emplaced once. In the second branch, it is emplaced n times, so the shared pointer needs to be copied (which is unfortunately implicit in C++..). This part is the reason why I used shared pointers, to avoid copying the JSON configs too often.